src/billing_handler.rs

```diff
@@ -1,33 +1,53 @@
 use crate::model::AppState;
 
-use actix_web::{web::Data, HttpResponse};
+use actix_web::web::Data;
+use actix_web::HttpResponse;
 use anyhow::Context;
 use k256::elliptic_curve::generic_array::sequence::Lengthen;
 use serde_json::json;
 use tiny_keccak::{Hasher, Keccak};
 
 pub async fn billing_data(appstate: Data<AppState>) -> HttpResponse {
-    let mut costs_gaurd = appstate.execution_costs.lock().unwrap();
-    let costs_map = costs_gaurd.clone();
+    let mut execution_costs_gaurd = appstate.execution_costs.lock().unwrap();
+
+    if execution_costs_gaurd.is_empty() {
+        return HttpResponse::BadRequest().body("No billing data available");
+    }
+
+    let mut billing_data: Vec<u8> = Vec::new();
+
+    for (tx_hash, cost) in execution_costs_gaurd.iter() {
+        let mut bytes32_tx_hash = [0u8; 32];
+        if let Err(err) = hex::decode_to_slice(&tx_hash[2..], &mut bytes32_tx_hash) {
+            return HttpResponse::InternalServerError()
+                .body(format!("Error decoding transaction hash: {:?}", err));
+        }
+        
+        billing_data.append(&mut bytes32_tx_hash.to_vec());
+        billing_data.append(&mut cost.to_be_bytes().to_vec());
+    }
+
+    let mut hasher = Keccak::v256();
+    hasher.update(&billing_data);
 
     let mut hash = [0u8; 32];
-    let mut hasher_gaurd = appstate.billing_hasher.lock().unwrap();
-    hasher_gaurd.clone().finalize(&mut hash);
+    hasher.finalize(&mut hash);
+
     let sign = appstate
         .signer
         .sign_prehash_recoverable(&hash)
-        .context("Failed to sign requests data");
+        .context("Failed to sign billing data");
     if sign.is_err() {
-        return HttpResponse::InternalServerError().body(format!("{:?}", sign.unwrap_err()));
+        return HttpResponse::InternalServerError().body(format!("{}", sign.unwrap_err()));
     }
     let (rs, v) = sign.unwrap();
     let signature = hex::encode(rs.to_bytes().append(27 + v.to_byte()).as_slice());
 
-    costs_gaurd.clear();
-    hasher_gaurd.clone_from(&Keccak::v256());
+    let billing_data = hex::encode(billing_data.as_slice());
+    execution_costs_gaurd.clear();
```

Comment:
It's risky to clean up execution_costs here. Consider a case where HTTP connection got disconnected just before client can receive the costs. And these costs will be lost forever. So, we need some mechanism that costs only get deleted once client has confirmed the receipt of costs.

One idea is that client will submit the transaction and give txn hash as receipt and then we clear the costs. What do you think?

Reply:
Even if the client (program running outside the enclave) just receives the data it'll be still fine because it'll be stored there until they make the transaction. The thing is how do we want to receive the confirmation from the client. One way I can think of right now is to have a scheduler running in background which will interact with the client on regular intervals to receive the confirmation. Can you suggest something better?

Reply:
Then, add another endpoint `ClearSettledBills. Client call this endpoint with list of hashes as arguments, and handler will just cleans up old bills corresponding to those hashes. So, you would need to maintain another object which will hold the generated bills.

Reply:
this won't really work, we cannot trust the user to tell us what has been billed and what hasn't since he can then use it to double dip

Reply:
What exactly is double dip? Is it posting the generated bill more than 1 time on chain?